src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
```

Comment:
> I don't understand why foo.d's object code should include helper functions for bar.d.

@WalterBright Here *was* the problematic part. If `global.params.multiobj == false`, it's not a problem.
But when `global.params.multiobj == true`, in the library compilation:

```
all1_order_flipped:
	${dmd} -oflibproj6.a -lib -g project2.d foo2.d foo1.d
```

Both `foo2` and `foo1` imports `stdio`, then `stdio`->importedFrom` is set to `foo2`. And this code has emit the `stdio` module helpers into the libraried `foo2.obj`, which also contains `foo2.__ModuleInfoZ`.